content/blog/2020-12-19-asmm/index.md

```diff
@@ -0,0 +1,119 @@
+
+With the exception of the ```mat-mul.bril``` benchmark, these benchmarks are not a very difficult evaluation. It may just be that full flexibility in memory management is hard to keep track of anyways so users are already restricting their use and these benchmarks are representative of some standard use. To compensate for this, I've gone and included the interpreter memory test cases to get fuller coverage of some of the edge cases that could occur. For instance, ```alloc_many.bril``` is a go example of a program that would fail to check in this bril extension.
+
+Alloc_many.bril
+
+```
+@main {
+  v: int = const 1;
+  max: int = const 1000000;
+  count: int = const 0;
+.lbl:
+  count: int = add count v;
+  p: ptr<int> = alloc v;
+  free p;
+  loop: bool = ge count max;
+  br loop .end .lbl;
+.end:
+  print count;
+}
+
+```
+
+Notice how this program loops over the loop body some `max` number of times. Each time, the program allocates and frees some memory. Because the same variable destination is being used each time, this would not check. The allocation would need to be moved to the loop header and either freed immediately or at block `.end`. If there were user defined regions that you could write something like:
+
+Alloc_many.bril.asmm
+
+```
+@main {
+  v: int = const 1;
+  max: int = const 1000000;
+  count: int = const 0;
+.lbl:
+  start_region "loop_body"
+  count: int = add count v;
+  p: ptr<int> = alloc v;
+  loop: bool = ge count max;
+  end_region "loop_body"
+  br loop .end .lbl;
+.end:
+  print count;
+}
+```
+
+### Correctness
+
+The correctness argument comes in two parts. First by using all of the memory tests/benchmarks, this bril extension will have as much coverage/correctness as the interpreter itself. The more logical argument some from the limitations that have been enforced that remove complex control flow like branches and loops for ```alloc``` calls. This means that you can look at the path of the dominator tree from the root to exit leaf nodes to find all of the owner pointers created for this region. Also by not including user created regions, this simplifies the region hierarchy with the current function at the bottom and each of it's arguments has a region different from the function one that is above it in the hierarchy. Another important part around function calls is that function arguments are always of a borrowed pointer type and the return pointer is always an owned pointer type. Therefore, you cannot duplicate a function pointer on accident and try ot free it twice. This gives a clear divide, it is the callers job to do the memory management for any pointer it gives or receives from the callee. It is a bit unclear if a programmer could give purposely wrong annotations to violate the soundness of this transformation. 
```

Comment:
> It is a bit unclear if a programmer could give purposely wrong annotations to violate the soundness of this transformation.

That's pretty worrisome! :smiley: Some clarity on what you mean by this would be helpful…

In particular, does your compiler actually enforce that returned values are owned pointers? And presumably it doesn't free them?

Reply:
Returned values are freed by the caller, not the callee and must be owned so that part is enforced. I will provide an example that I'm still unsure with how to handle that motivates this worry. In general though, I've tried to design the extension to be straightforward to infer. I would much rather a programmer just rely on the inference step to handle annotations rather than writing them directly.